### PR TITLE
Add OneAgentIM CRD

### PR DESCRIPTION
**Pull request by DTMad**

### CODE REVIEW

File: pkg/controller/oneagent/oneagent_controller.go [hunk 1]
```diff
@@ -344,19 +347,19 @@ func (r *ReconcileOneAgent) reconcileVersion(logger logr.Logger, instance *dynat
 
 	// Workaround: 'instances' can be null, making DeepEqual() return false when comparing against an empty map instance.
 	// So, compare as long there is data.
-	if (len(instances) > 0 || len(instance.Status.Instances) > 0) && !reflect.DeepEqual(instances, instance.Status.Instances) {
-		logger.Info("oneagent pod instances changed", "status", instance.Status)
+	if (len(instances) > 0 || len(instance.GetOneAgentStatus().Instances) > 0) && !reflect.DeepEqual(instances, instance.GetOneAgentStatus().Instances) {
+		logger.Info("oneagent pod instances changed", "status", instance.GetOneAgentStatus)
```
Comment (lrgar) [inline pkg/controller/oneagent/oneagent_controller.go:351]:
```suggestion
		logger.Info("oneagent pod instances changed", "status", instance.GetOneAgentStatus())
```

File: pkg/controller/oneagentim/oneagentim_controller.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
+/*
+Copyright YEAR Dynatrace LLC.
```
Comment (lrgar) [inline pkg/controller/oneagentim/oneagentim_controller.go:2]:
```suggestion
Copyright 2020 Dynatrace LLC.
```

File: pkg/controller/add_oneagentim.go [hunk 3]
```diff
@@ -0,0 +1,22 @@
+/*
+Copyright YEAR Dynatrace LLC.
```
Comment (lrgar) [inline pkg/controller/add_oneagentim.go:2]:
```suggestion
Copyright 2020 Dynatrace LLC.
```

File: pkg/apis/dynatrace/v1alpha1/oneagentim_types.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+// OneAgentIM configures the Dynatrace OneAgent for application monitoring
+// +k8s:openapi-gen=true
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:path=OneAgentIMs,scope=Namespaced,categories=dynatrace
+// +kubebuilder:printcolumn:name="ApiUrl",type=string,JSONPath=`.spec.apiUrl`
+// +kubebuilder:printcolumn:name="Tokens",type=string,JSONPath=`.spec.tokens`
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+// +operator-sdk:gen-csv:customresourcedefinitions.displayName="Dynatrace OneAgent Application Monitoring"
```
Comment (lrgar) [inline pkg/apis/dynatrace/v1alpha1/oneagentim_types.go:31]:
```suggestion
// +operator-sdk:gen-csv:customresourcedefinitions.displayName="Dynatrace OneAgent Infrastructure Monitoring"
```

File: pkg/apis/dynatrace/v1alpha1/oneagentim_types.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// OneAgentIM configures the Dynatrace OneAgent for application monitoring
```
Comment (lrgar) [inline pkg/apis/dynatrace/v1alpha1/oneagentim_types.go:23]:
```suggestion
// OneAgentIM configures the Dynatrace OneAgent for infrastructure monitoring
```

File: pkg/apis/dynatrace/v1alpha1/oneagentim_types.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+// +k8s:openapi-gen=true
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:path=OneAgentIMs,scope=Namespaced,categories=dynatrace
+// +kubebuilder:printcolumn:name="ApiUrl",type=string,JSONPath=`.spec.apiUrl`
+// +kubebuilder:printcolumn:name="Tokens",type=string,JSONPath=`.spec.tokens`
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
```
Comment (lrgar) [inline pkg/apis/dynatrace/v1alpha1/oneagentim_types.go:30]: I suggest adding the Phase column as we have on the OneAgent CRD.

File: pkg/controller/oneagentim/oneagentim_controller_test.go [hunk 7]
```diff
@@ -0,0 +1 @@
+package oneagentim
```
Comment (lrgar) [inline pkg/controller/oneagentim/oneagentim_controller_test.go:1]: If we're not doing anything, should we have this file?

Comment (DTMad) [reply pkg/controller/oneagentim/oneagentim_controller_test.go:1]: Will delete it

File: pkg/controller/oneagent/oneagent_controller.go [hunk 8]
```diff
@@ -391,25 +394,29 @@ func newDaemonSetForCR(instance *dynatracev1alpha1.OneAgent) *appsv1.DaemonSet {
...
-	args := instance.Spec.Args
-	if instance.Spec.Proxy != nil && (instance.Spec.Proxy.ValueFrom != "" || instance.Spec.Proxy.Value != "") {
-		args = append(instance.Spec.Args, "--set-proxy=$(https_proxy)")
+	args := instance.GetOneAgentSpec().Args
+	if instance.GetOneAgentSpec().Proxy != nil && (instance.GetOneAgentSpec().Proxy.ValueFrom != "" || instance.GetOneAgentSpec().Proxy.Value != "") {
+		args = append(instance.GetOneAgentSpec().Args, "--set-proxy=$(https_proxy)")
+	}
+
+	if _, ok := instance.(*dynatracev1alpha1.OneAgentIM); ok {
+		args = append(instance.GetOneAgentSpec().Args, "--set-infra-only=true")
```
Comment (lrgar) [inline pkg/controller/oneagent/oneagent_controller.go:419]: You don't want to override the proxy if set as well.

```suggestion
		args = append(args, "--set-infra-only=true")
```

**Pull request merged**